modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/containers/Form.es.js

```diff
@@ -0,0 +1,206 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import dom from 'metal-dom';
+import Soy from 'metal-soy';
+import React, {
+	useCallback,
+	useEffect,
+	useImperativeHandle,
+	useRef,
+} from 'react';
+
+import Pages from '../components/Pages.es';
+import {FormProvider, useForm} from '../hooks/useForm.es';
+import formValidate from '../thunks/formValidate.es';
+import {getConnectedReactComponentAdapter} from '../util/ReactComponentAdapter.es';
+import {evaluate} from '../util/evaluation.es';
+import {getFormId, getFormNode} from '../util/formId.es';
+import templates from './Form.soy';
+
+function handleLiferayFormSubmitted(event) {
+	if (event.form && event.form.getDOM() === this.form) {
+		event.preventDefault();
+	}
+}
+
+const Form = React.forwardRef((props, ref) => {
+	const dispatch = useForm();
+	const containerRef = useRef(null);
+
+	const validate = useCallback(() => {
+		const {
+			activePage,
+			defaultLanguageId,
+			editingLanguageId,
+			pages,
+			portletNamespace,
+			rules,
+		} = props;
+
+		return dispatch(
+			formValidate({
+				activePage,
+				defaultLanguageId,
+				editingLanguageId,
+				pages,
+				portletNamespace,
+				rules,
+			})
+		);
+	}, [dispatch, props]);
```

Comment:
Yeah, I kept thinking about it a bit, in the end, it will always generate a new object due to the `pages`, so I'm wondering if I remove this completely.